embassy-executor/src/raw/deadline.rs

```diff
@@ -0,0 +1,156 @@
+use core::future::{poll_fn, Future};
+use core::task::Poll;
+
+/// A type for interacting with the deadline of the current task
+///
+/// Requires the `edf-scheduler` feature
+pub struct Deadline {
+    /// Deadline value in ticks, same time base and ticks as `embassy-time`
+    pub instant_ticks: u64,
+}
+
+impl Deadline {
+    /// Sentinel value representing an "unset" deadline, which has lower priority
+    /// than any other set deadline value
+    pub const UNSET_DEADLINE_TICKS: u64 = u64::MAX;
+
+    /// Does the given Deadline represent an "unset" deadline?
+    #[inline]
+    pub fn is_unset(&self) -> bool {
+        self.instant_ticks == Self::UNSET_DEADLINE_TICKS
+    }
+
+    /// Set the current task's deadline at exactly `instant_ticks`
+    ///
+    /// This method is a future in order to access the currently executing task's
+    /// header which contains the deadline.
+    ///
+    /// Analogous to `Timer::at`.
+    ///
+    /// This method does NOT check whether the deadline has already passed.
+    #[must_use = "Setting deadline must be polled to be effective"]
+    pub fn set_current_task_deadline(instant_ticks: u64) -> impl Future<Output = ()> {
+        poll_fn(move |cx| {
+            let task = super::task_from_waker(cx.waker());
+            // SAFETY: A task can only modify its own deadline, while the task is being
+            // polled, meaning that there cannot be concurrent access to the deadline.
```

Comment:
this is unfortunately unsound. A user can do:

- Create a task on thread A
- Get the task polled
- Take the waker
- Move that waker to Thread B
- on Thread B, construct a new Context from that waker
- call `set_current_task_deadline`. 

this causes a data race between thread A and B.

Reply:
I'm not sure if I follow, this executes in the context of a single poll by a single executor, you can't have two threads touching a single TaskRef at once, right?

Reply:
I mean there's no guarantee that `super::task_from_waker(cx.waker())` is the task currently being polled. 

You can move a task's waker to any thread, create a context and call `set_current_task_deadline` with it.

Reply:
I think there's two ways out:

- Store it as `Mutex<Cell<u64>>`, take a CS.
- Store it as 2x `AtomicU32`. If the user does the above "waker smuggling" they could cause the executor to see a "torn" deadline value, but it's not UB anymore, it's just "wrong behavior".

Reply:
Right, the solution to `Spawner::for_current_executor` was to make it unsafe.
But for this API that'd be very unfortunate.

I think going with the two atomics is nicest.
@jamesmunns if you give me access I can make the changes.

Reply:
@diondokter added you as a collaborator to my fork.

Reply:
@Dirbaio This should now be addressed by the changes made in https://github.com/jamesmunns/embassy/pull/2.